# 亚马逊
## 简介：该数据集包含超过一千条亚马逊产品的评分、评论、价格等信息，数据来自于亚马逊官方网站上列出的详细信息。
变量含义：
- product_id ：产品id
- product_name：产品的名称
- category：产品的分类
- discounted_price：产品的折扣价格
- actual_price：产品的实际价格
- discount_percentage：产品折扣的百分比
- rating：消费者的评分
- rating_count：参与评分的消费者数量
- about_product：产品的描述
- user_id：为产品撰写评论的消费者id
- user_name：为产品撰写评论的消费者名称
- review_id：评论的id
- review_title：简短的评论题目
- review_content：评论内容
- img_link：产品的图片链接
- product_link：产品的官网链接

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv("amazon.csv")
original_data.head()

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


## 评估数据整齐度

In [3]:
original_data.sample(10)

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
862,B09NC2TY11,Noise ColorFit Ultra Buzz Bluetooth Calling Sm...,Electronics|WearableTechnology|SmartWatches,"₹2,499","₹5,999",58%,4.1,"5,852",Bluetooth calling: Stay connected with the wor...,"AHI7MFQMUC5N6DRDGTYZPAKPWF2A,AF4YTGSRB6V6GKOAS...","Sasmita Chopdar,Manas Mishra,Atul kumar,Rudra ...","R3K08458ILZK0F,R3OJTSZV57IWTC,R1DLM3QOLR43NS,R...","Good quality,Bindaas watch,Good,Must buy,The L...","Taking to school or exercise,Must buy it if u'...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Noise-ColorFit-Bluetooth...
1399,B088WCFPQF,Cafe JEI French Press Coffee and Tea Maker 600...,"Home&Kitchen|Kitchen&HomeAppliances|Coffee,Tea...","₹1,099","₹1,500",27%,4.5,"1,065",4 LEVEL FILTRATION SYSTEM - We use double stai...,"AFRB32NPLQW24ZGJTXRYK6OUI2HA,AG2PGVWTVLRVZRJCJ...","Deepak,Garuda,Shikha,Rajesh Manoharan,Gaurav,C...","R3EFB0EG66OLOX,RIND9MF93GBO0,R38ISD2RSB4M70,R3...","Absolutely loving it!,Excellent product,Very n...",I gifted it to my mother last month; it was li...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Cafe-JEI-Filtration-Resi...
752,B09M869Z5V,Portronics MPORT 31C 4-in-1 USB Hub (Type C to...,Computers&Accessories|Accessories&Peripherals|...,₹570,₹999,43%,4.2,"3,201","4-in-1 Type-C USB Hub, 1 3.0 USB-A port, and 3...","AG7L3FBDA2KLNZOZWS2XF7Y2GGRQ,AH6VYUL52LC2OCCRB...","A K Rai,Poojitha Subrahmanyam Naralasetti,Mugi...","R26P3IBAM6K3G2,R3CVDCIJEXR401,RD43FPKWA79MH,R1...","Value For Money But....,Compact and Budget fri...",It is very good. Product Quality is decent but...,https://m.media-amazon.com/images/I/21qdAZyu9x...,https://www.amazon.in/Portronics-MPORT-Type-Po...
1147,B07WKB69RS,Lifelong LLWH106 Flash 3 Litres Instant Water ...,"Home&Kitchen|Heating,Cooling&AirQuality|WaterH...","₹2,088","₹5,550",62%,4,"5,292",8 bar pressure rated: High pressure withstandi...,"AHODVRQWWJ6ZANKRQMUTC2XAP7DA,AHUK7JORZ6JUBP6Y2...","Shreya Saxena,Asish Kumar Saha,Nikhil Kumar,Di...","R36G8V9B8EIG4Z,R1UQJ38MFDF636,R3GHKCA6I36EBF,R...",No Inlet and Outlet Pipe + Installation is Cha...,Pros:1. Most affordable option2. 2 year warran...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Lifelong-Flash-Instant-H...
800,B07JF9B592,MAONO AU-400 Lavalier Auxiliary Omnidirectiona...,MusicalInstruments|Microphones|Condenser,₹478,₹699,32%,3.8,"20,218",4-pole jack is compatible for laptop which has...,"AHQ4Q75NBEWOM4OWOXUZW7V247NQ,AEQCAMSZJTMNIKXAP...","ˢᴰ82ˢᶜ ✔️,Manish Rautela,Sai kiran,Dinesh S,Ut...","RKBKQKSEET7CC,RQM34GWJC0DPJ,R1PJNXT9PME2I1,R2V...","loud & clear,Avrage in budget range,Best for b...",at Rs.319 it's a great deal. I use it for mobi...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Maono-AU-400-Lavalier-Mi...
455,B09JS562TP,Motorola a10 Dual Sim keypad Mobile with 1750 ...,Electronics|Mobiles&Accessories|Smartphones&Ba...,"₹1,399","₹1,630",14%,4.0,"9,378",Ensured Power packed performance with Mediatek...,"AEN657OFUBBVTAFRFCOOUKFBNQ4Q,AESZZZXVFKLKXWSQP...","Rajasekhar,Dimpal,rajat kumar,akanksha,Noone,H...","R27C4TPKHXYBRU,R1WGISGIIXAU1B,R2WFSJJW04UWJ8,R...",Out of 5 iam giving 3.5 rating everything is o...,rating everything is okay except voice sound d...,https://m.media-amazon.com/images/I/41MOWVL2YN...,https://www.amazon.in/Motorola-keypad-Mobile-E...
201,B00GGGOYEK,Storite USB 2.0 A to Mini 5 pin B Cable for Ex...,Computers&Accessories|Accessories&Peripherals|...,₹259,₹699,63%,3.8,"2,399",Make plug-and-play connections with mobile pho...,"AFHX6LN2EGRSLCIKZERTK236KJWA,AGM6VKOVQWLVZW5NX...","Angsuman DUtta,Amjath sp,Amazon Customer,utpal...","R7CW64V48YJHE,R185CPLU005RPS,R2R70NKW75DZAS,R3...","Useful but the length is a bit short,This prod...",Product is good but the length of the wire is ...,https://m.media-amazon.com/images/W/WEB

从抽样的10行数据数据来看，数据不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。 需要调整的结构性问题为 category ，应该需要扩展为多个单元格，而非一个单元格

## 数据结构性调整

In [4]:
original_data[["category1","category2","category3","category4","category5","category6","category7"]]=original_data["category"].str.split("|",expand = True)
original_data
#这一步去掉了category原有的冗长内容

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,...,review_content,img_link,product_link,category1,category2,category3,category4,category5,category6,category7
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...",...,Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...",...,I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...",...,"Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...",...,"Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...",...,"Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...,₹379,₹919,59%,4,"1,090",SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...,"AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...",...,"I received product without spanner,Excellent p...",https://m.media-amazon.com/images/I/41fDdRtjfx...,https://www.amazon.in/Noir-Aqua-Spanner-Purifi...,Home&Kitchen,Kitchen&HomeAppliances,WaterPurifiers&Accessories,WaterPurifierAccessories,None,None,None
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...,"₹2,280","₹3,045",25%,4.1,"4,118","230 Volts, 400 watts, 1 Year","AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...",...,"ok,got everything as mentioned but the measuri...",https://m.media-amazon.com/images/I/41gzDxk4+k...,https://www.amazon.in/Prestige-Delight-PRWO-1-...,Home&Kitchen,Kitchen&HomeAppliances,SmallKitchenAppliances,Rice&PastaCookers,None,None,None
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,"Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...","₹2,219","₹3,080",28%,3.6,468,International design and styling|Two heat sett...,"AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY

In [5]:
original_data=original_data.drop("category",axis=1)
original_data

,product_id,product_name,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,...,review_content,img_link,product_link,category1,category2,category3,category4,category5,category6,category7
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...",...,Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...",...,I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...",...,"Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...",...,"Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...",...,"Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1460,B08L7J3T31,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,₹379,₹919,59%,4,"1,090",SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...,"AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...","Prabha ds,Raghuram bk,Real Deal,Amazon Custome...",...,"I received product without spanner,Excellent p...",https://m.media-amazon.com/images/I/41fDdRtjfx...,https://www.amazon.in/Noir-Aqua-Spanner-Purifi...,Home&Kitchen,Kitchen&HomeAppliances,WaterPurifiers&Accessories,WaterPurifierAccessories,None,None,None
1461,B01M6453MB,Prestige Delight PRWO Electric Rice Cooker (1 ...,"₹2,280","₹3,045",25%,4.1,"4,118","230 Volts, 400 watts, 1 Year","AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...","Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...",...,"ok,got everything as mentioned but the measuri...",https://m.media-amazon.com/images/I/41gzDxk4+k...,https://www.amazon.in/Prestige-Delight-PRWO-1-...,Home&Kitchen,Kitchen&HomeAppliances,SmallKitchenAppliances,Rice&PastaCookers,None,None,None
1462,B009P2LIL4,Bajaj Majesty RX10 2000 Watts Heat Convector R...,"₹2,219","₹3,080",28%,3.6,468,International design and styling|Two heat sett...,"AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...","Nehal Desai,Danish Parwe

In [6]:
#产品名称这里难以归类 所以不改变

## 评估数据干净度

In [7]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   discounted_price     1465 non-null   object
 3   actual_price         1465 non-null   object
 4   discount_percentage  1465 non-null   object
 5   rating               1465 non-null   object
 6   rating_count         1463 non-null   object
 7   about_product        1465 non-null   object
 8   user_id              1465 non-null   object
 9   user_name            1465 non-null   object
 10  review_id            1465 non-null   object
 11  review_title         1465 non-null   object
 12  review_content       1465 non-null   object
 13  img_link             1465 non-null   object
 14  product_link         1465 non-null   object
 15  category1            1465 non-null   object
 16  catego

从输出结果来看，数据共有1465条观察值，而rating_count变量存在缺失值。

此外，discounted_price等的数据类型应为数值，应当进行数据格式转换。

## 评估缺失数据

In [8]:
original_data[original_data["rating_count"].isnull()]

,product_id,product_name,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,...,review_content,img_link,product_link,category1,category2,category3,category4,category5,category6,category7
282,B0B94JPY2N,Amazon Brand - Solimo 65W Fast Charging Braide...,₹199,₹999,80%,3.0,NaN,USB C to C Cable: This cable has type C connec...,AE7CFHY23VAJT2FI4NZKKP6GS2UQ,Pranav,...,I have a pd supported car charger and I bought...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Amazon-Brand-Charging-Su...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None
324,B0BQRJ3C47,"REDTECH USB-C to Lightning Cable 3.3FT, [Apple...",₹249,₹999,75%,5.0,NaN,💎[The Fastest Charge] - This iPhone USB C cabl...,AGJC5O5H5BBXWUV7WRIEIOOR3TVQ,Abdul Gafur,...,Quick delivery.Awesome ProductPacking was good...,https://m.media-amazon.com/images/I/31-q0xhaTA...,https://www.amazon.in/REDTECH-Lightning-Certif...,Computers&Accessories,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,None,None


可以看到rating_count的值为nan，说明没有人评分，但是其rating存在数值下
面存在两种可能，一是此评分为默认评分，二是数据出现问题，所以要验证review_id看看用书是否村子

In [9]:
original_data[original_data["review_id"].isnull()]

,product_id,product_name,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,...,review_content,img_link,product_link,category1,category2,category3,category4,category5,category6,category7


可以看到 不存在没有review_id的情况，所以应该删除上述数据

## 评估重复数据

根据数据变量的含义来看 针对此数据集，我们无需评估重复数据。

## 评估不一致数据

根据数据变量的含义来看 没有会产生不一致数据的变量

## 评估无效或错误数据

In [10]:
original_data.describe()

,product_id,product_name,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,...,review_content,img_link,product_link,category1,category2,category3,category4,category5,category6,category7
count,1465,1465,1465,1465,1465,1465,1463,1465,1465,1465,...,1465,1465,1465,1465,1465,1457,1300,522,85,13
unique,1351,1337,550,449,92,28,1143,1293,1194,1194,...,1212,1412,1465,9,29,71,136,56,14,7
top,B07JW9H4J1,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",₹199,₹999,50%,4.1,"9,378",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...","$@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...",...,"I am not big on camera usage, personally. I wa...",https://m.media-amazon.com/images/I/413sCRKobN...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,Electronics,Accessories&Peripherals,Cables&Accessories,Cables,USBCables,DryIrons,StickBallpointPens
freq,3,5,53,120,56,244,9,6,10,10,...,8,3,1,526,381,240,267,233,24,3


未发现无效或错误数据

# 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：
把discounted_price 变量的数据类型转换为为数值
把actual_price 变量的数据类型转换为数值
把discount_percentage变量的数据类型转换为数值
把rating 变量的数据类型转换为数值
把rating_count 变量的数据类型转换为数值
把rating变量缺失的观察值删除

In [11]:
cleaned_data= original_data.copy()

In [12]:
cleaned_data["discounted_price"] = cleaned_data["discounted_price"] .str.extract(r'([\d\.]+)')[0].astype(float)

In [13]:
cleaned_data["actual_price"] = cleaned_data["actual_price"] .str.extract(r'([\d\.]+)')[0].astype(float)

In [14]:
cleaned_data["discount_percentage"]  = cleaned_data["discount_percentage"] .str.replace('%', '', regex=True).astype(float) / 100

In [15]:
cleaned_data["rating_count"] = cleaned_data["rating_count"] .str.replace(',', '').astype(float)

In [16]:
cleaned_data["rating"] = pd.to_numeric(cleaned_data["rating"].str.replace(',', ''), errors='coerce')

In [18]:
cleaned_data=cleaned_data.rename(columns= {"discounted_price":"discounted_price(₹)","actual_price":"actual_price(₹)","discount_percentage":"discount_percentage(%)"})

In [19]:
cleaned_data=cleaned_data[cleaned_data["rating_count"]>0]

# 保存清理后的数据

In [20]:
cleaned_data.to_csv("amazon_cleaned.csv", index=False)